src/r_fiber.rs

```diff
@@ -0,0 +1,388 @@
+        write!(f, "{}", self.inspect())
+    }
+}
+
+impl RFiber {
+    /// Return `Some(RFiber)` if `val` is a `RFiber`, `None` otherwise.
+    #[inline]
+    pub fn from_value(val: Value) -> Option<Self> {
+        unsafe {
+            Value::new(rb_obj_is_fiber(val.as_rb_value()))
+                .to_bool()
+                .then(|| Self(NonZeroValue::new_unchecked(val)))
+        }
+    }
+}
+
+impl RFiber {
+    /**
+     * Fiber.new
+     */
+    pub fn blocking<F, R>(block: F) -> Self
+    where
+        F: 'static + FnMut(&[Value], Option<Proc>) -> R,
+        R: BlockReturn,
+    {
+        unsafe extern "C" fn call<F, R>(
+            _yielded_arg: VALUE,
+            callback_arg: VALUE,
+            argc: c_int,
+            argv: *const VALUE,
+            blockarg: VALUE,
+        ) -> VALUE
+        where
+            F: FnMut(&[Value], Option<Proc>) -> R,
+            R: BlockReturn,
+        {
+            let closure = &mut *(callback_arg as *mut F);
+            Block::new(closure)
+                .call_handle_error(argc, argv as *const Value, Value::new(blockarg))
+                .as_rb_value()
+        }
+
+        let (closure, keepalive) = wrap_closure(block);
+        let call_func =
+            call::<F, R> as unsafe extern "C" fn(VALUE, VALUE, c_int, *const VALUE, VALUE) -> VALUE;
+        #[cfg(ruby_lt_2_7)]
+        let call_func: unsafe extern "C" fn() -> VALUE = unsafe { std::mem::transmute(call_func) };
+
+        let fiber = Self(unsafe {
+            NonZeroValue::new_unchecked(Value::new(rb_fiber_new(Some(call_func), closure as VALUE)))
+        });
+        fiber.ivar_set("__rust_closure", keepalive).unwrap();
+        fiber
+    }
+
+    /**
+     * Fiber.new
+     */
+    pub fn non_blocking(block: Proc) -> Result<Self, Error> {
+        let fiber_class = define_class("Fiber", Default::default())?;
+        let fib = fiber_class.funcall_with_block_kw::<_, _, _, Value>(
+            "new",
+            (),
+            (("blocking", false),),
+            block,
+        )?;
+        Ok(Self(unsafe { NonZeroValue::new_unchecked(fib) }))
+    }
+
+    /**
+     * Fiber.current
+     */
+    pub fn current() -> Self {
+        Self(unsafe { NonZeroValue::new_unchecked(Value::new(rb_fiber_current())) })
+    }
+
+    /**
+     * Fiber.yield
+     */
+    pub fn rb_yield<A>(args: A) -> Result<Value, Error>
+    where
+        A: ArgList,
+    {
+        let args = args.into_arg_list();
+        let args = args.as_ref();
+        protect(|| {
+            Value::new(unsafe {
+                rb_fiber_yield(args.len() as c_int, args.as_ptr() as *const VALUE)
+            })
+        })
+    }
+
+    /**
+     * fiber.alive?
+     */
+    pub fn alive(&self) -> bool {
+        unsafe { Value::new(rb_fiber_alive_p(self.as_rb_value())) }.to_bool()
+    }
+
+    /**
+     * fiber.resume
+     */
+    pub fn resume<A>(&self, args: A) -> Result<Value, Error>
+    where
+        A: ArgList,
+    {
```

Comment:
Can this call `try_convert()`? Like:
```rust
pub fn resume<A, T>(&self, args: A) -> Result<T, Error>
where
    A: ArgList,
    T: TryConvert,
{
    ...
    .and_then(|v| v.try_convert())
}
```

Reply:
done. and added it for all the other methods.

One thing I also added in relation to this is `impl TryConvert for ()`, because of ergonomics.

Right now this checks for `is_nil`, but not sure if thats necessary or not, you call.